core/src/main/scala/kafka/controller/ReplicaStateMachineV2.scala

```diff
@@ -0,0 +1,316 @@
+    var results = Map.empty[TopicAndPartition, LeaderIsrAndControllerEpoch]
+    var remaining = partitions
+    while (remaining.nonEmpty) {
+      val (successfulRemovals, removalsToRetry) = doRemoveReplicasFromIsr(replicaId, remaining)
+      results ++= successfulRemovals
+      remaining = removalsToRetry
+    }
+    results
+  }
+
+  private def doRemoveReplicasFromIsr(replicaId: Int, partitions: Seq[TopicAndPartition]) = {
+    val (leaderAndIsrs, partitionsWithNoLeaderAndIsrInZk, failedStateReads) = getTopicPartitionStatesFromZk(replicaId, partitions)
+    val (leaderAndIsrsWithReplica, leaderAndIsrsWithoutReplica) = leaderAndIsrs.partition { case (partition, leaderAndIsr) => leaderAndIsr.isr.contains(replicaId) }
+    val (leaderAndIsrsRiskingEmptyIsr, leaderAndIsrsNotRiskingEmptyIsr) = leaderAndIsrsWithReplica.partition { case (partition, leaderAndIsr) =>
+      leaderAndIsr.isr.size == 1 &&  leaderAndIsr.isr.head == replicaId
+    }
+    val (adjustedLeaderAndIsrsForPartitionsWithEmptyIsr, failedLogConfigReads) = leaderAndIsrBasedOnLogConfigs(replicaId, leaderAndIsrsRiskingEmptyIsr)
+    val adjustedLeaderAndIsrs = leaderAndIsrsNotRiskingEmptyIsr.mapValues { leaderAndIsr =>
+      val newLeader = if (replicaId == leaderAndIsr.leader) LeaderAndIsr.NoLeader else leaderAndIsr.leader
+      val adjustedIsr = leaderAndIsr.isr.filter(_ != replicaId)
+      leaderAndIsr.newLeaderAndIsr(newLeader, adjustedIsr)
+    } ++ adjustedLeaderAndIsrsForPartitionsWithEmptyIsr
+    val (successfulUpdates, updatesToRetry, failedUpdates) = zkUtils.updateLeaderAndIsr(adjustedLeaderAndIsrs, controller.epoch)
+    partitionsWithNoLeaderAndIsrInZk.foreach { partition =>
+      if (!controller.topicDeletionManager.isPartitionToBeDeleted(partition)) {
+        val replica = PartitionAndReplica(partition.topic, partition.partition, replicaId)
+        val exception = new StateChangeFailedException(s"Failed to change state of replica $replicaId for partition $partition since the leader and isr path in zookeeper is empty")
+        logFailedStateChange(replica, replicaState(replica), OfflineReplica, exception)
+      }
+    }
+    (failedStateReads ++ failedLogConfigReads ++ failedUpdates).foreach { case (partition, e) =>
+      val replica = PartitionAndReplica(partition.topic, partition.partition, replicaId)
+      logFailedStateChange(replica, replicaState(replica), OfflineReplica, e)
+    }
+    val leaderIsrAndControllerEpochs = (leaderAndIsrsWithoutReplica ++ successfulUpdates).map { case (partition, leaderAndIsr) =>
+      val leaderIsrAndControllerEpoch = LeaderIsrAndControllerEpoch(leaderAndIsr, controller.epoch)
+      controllerContext.partitionLeadershipInfo.put(partition, leaderIsrAndControllerEpoch)
+      partition -> leaderIsrAndControllerEpoch
+    }
+    (leaderIsrAndControllerEpochs, updatesToRetry)
+  }
+
+  // replicas not in isr are implied as partitions in currentLeaderAndIsrs that are not in candidateLeaderAndIsrs
```

Comment:
Hmm, not sure that I really understand this comment. Also, could we document the return value?